sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs

```diff
@@ -737,58 +680,24 @@ private Block BuildSwitchStatement_CSharp7(SwitchStatementSyntax switchStatement
-                            // Creating BranchBlock instead of BinaryBranchBlock will generate some
-                            // false negatives because we will not be able to evaluate the case label
-                            // condition. BinaryBranchBlock causes exceptions in the CSharpExplodedGraph
-                            // because there is no instruction for the branch condition. Such instruction
-                            // would be difficult to generate because it is not present in the AST.
-                            // See https://github.com/SonarSource/sonar-csharp/issues/1801
-                            // and https://github.com/SonarSource/sonar-csharp/issues/1831
-
-                            currentSectionBlock = CreateBranchBlock(casePatternSwitchLabel,
-                                successors: new[] { sectionBlock, currentSectionBlock });
-                        }
-                        else
-                        {
                             currentSectionBlock = BuildCasePattern(casePatternSwitchLabel,
                                 trueSuccessor: sectionBlock, falseSuccessor: currentSectionBlock);
-                        }
-                    }
-                    else if (label is CaseSwitchLabelSyntax caseSwitchLabel &&
-                        caseSwitchLabel.Value != null &&
-                        caseSwitchLabel.Value.IsKind(SyntaxKind.NullLiteralExpression))
-                    {
-                        // Creating BranchBlock instead of BinaryBranchBlock will generate some
-                        // false negatives because we will not be able to evaluate the case label
-                        // condition. BinaryBranchBlock causes exceptions in the CSharpExplodedGraph
-                        // because there is no instruction for the branch condition. Such instruction
-                        // would be difficult to generate because it is not present in the AST.
-                        // See https://github.com/SonarSource/sonar-csharp/issues/1801
-
-                        currentSectionBlock = CreateBranchBlock(caseSwitchLabel,
-                             successors: new[] { sectionBlock, currentSectionBlock });
                     }
                     else if (label is CaseSwitchLabelSyntax simpleCaseLabel)
                     {
-                        currentSectionBlock = CreateBinaryBranchBlock(simpleCaseLabel.Value,
-                           trueSuccessor: sectionBlock,
-                           falseSuccessor: currentSectionBlock);
+                        currentSectionBlock = BuildExpression(switchStatement.Expression, CreateBinaryBranchBlock(simpleCaseLabel,
+                            sectionBlock, currentSectionBlock));
+                        var key = GetCaseIndexer(simpleCaseLabel.Value);
+                        caseBlocksByValue[key] = sectionBlock;
                     }
                 }
             }
 
             BreakTarget.Pop();
-

```

Comment:
[nitpick] please revert removing the empty lines - it adds some noise to the diff